### PR TITLE
[Feat] Add streaming to chatbot

### PR DESCRIPTION
**Pull request by rahul-tuli**
This PR adds streaming mode to `deepsparse.infer` chatbot

Test Command:
```bash
deepsparse.infer /home/rahul/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-base/deployment \
    --stream 
```

Output:
```bash
2023-09-26 11:56:38 deepsparse.transformers WARNING  The neuralmagic fork of transformers may not be installed. It can be installed via `pip install nm_transformers`
Using pad_token, but it is not set yet.
2023-09-26 11:56:41 deepsparse.transformers.pipelines.text_generation INFO     Compiling an auxiliary engine to process a prompt with a larger processing length. This improves performance, but may result in additional memory consumption.
2023-09-26 11:56:42 deepsparse.utils.onnx INFO     Overwriting in-place the input shapes of the transformer model at /home/rahul/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-base/deployment/model.onnx
DeepSparse, Copyright 2021-present / Neuralmagic, Inc. version: 1.6.0.20230910 COMMUNITY | (c66b57da) (release) (optimized) (system=avx512, binary=avx512)
2023-09-26 11:56:51 deepsparse.utils.onnx INFO     Overwriting in-place the input shapes of the transformer model at /home/rahul/.cache/sparsezoo/neuralmagic/codegen_mono-350m-bigpython_bigquery_thepile-base/deployment/model.onnx
User: def fact(
Bot: n):
    if n == 0:
        return 1
    else:
        return n * fact(n-1)

def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n-1)

def factorial_recursive(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive(n-1)

def factorial_recursive_2(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_2(n-1)

def factorial_recursive_3(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_3(n-1)

def factorial_recursive_4(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_4(n-1)

def factorial_recursive_5(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_5(n-1)

def factorial_recursive_6(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_6(n-1)

def factorial_recursive_7(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_7(n-1)

def factorial_recursive_8(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_8(n-1)

def factorial_recursive_9(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_9(n-1)

def factorial_recursive_10(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_10(n-1)

def factorial_recursive_11(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_11(n-1)

def factorial_recursive_12(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_12(n-1)

def factorial_recursive_13(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_13(n-1)

def factorial_recursive_14(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_14(n-1)

def factorial_recursive_15(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_15(n-1)

def factorial_recursive_16(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_16(n-1)

def factorial_recursive_17(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_17(n-1)

def factorial_recursive_18(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_18(n-1)

def factorial_recursive_19(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_19(n-1)

def factorial_recursive_20(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_20(n-1)

def factorial_recursive_21(n):
    if n == 0:
        return 1
    else:
        return n * factorial_recursive_21(n-1)

def factorial_recursive_22(n):
    if n == 0:
```

### CODE REVIEW

File: src/deepsparse/transformers/pipelines/chat.py [hunk 1]
```diff
@@ -146,6 +147,12 @@ def process_engine_outputs(
             engine_outputs, **kwargs
         )
         # create the ChatOutput from the data provided
+        if inspect.isgenerator(text_generation_output):
+            return (
```
Comment (rahul-tuli) [inline src/deepsparse/transformers/pipelines/chat.py:151]: creates a new generator from existing!

File: src/deepsparse/transformers/pipelines/chat.py [hunk 2]
```diff
@@ -146,6 +147,12 @@ def process_engine_outputs(
...
+                ChatOutput(**output.dict(), session_ids=session_ids)
+                for output in text_generation_output
+            )
```
Comment (rahul-tuli) [inline src/deepsparse/transformers/pipelines/chat.py:154]: creates a new generator from existing!

Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/chat.py:154]: 🏹

**Review by dbogunowicz — DISMISSED** [review]

File: src/deepsparse/transformers/infer.py [hunk 3]
```diff
@@ -131,32 +145,50 @@ def main(
...
+    )
+
+
+def _display_bot_response(stream: bool, response):
+    # print response from pipeline, streaming or not
+
+    print("Bot:", end=" ")
+    if stream:
+        for generation in response:
+            print(generation.generations[0].text, end=" ")
```
Comment (mgoin) [inline src/deepsparse/transformers/infer.py:187]:
```suggestion
    print("Bot:", end="")
    if stream:
        for generation in response:
            print(generation.generations[0].text, end="")
```

**Review by mgoin — APPROVED** [review]

File: src/deepsparse/transformers/inference/infer.py [hunk 4]
```diff
@@ -161,57 +174,74 @@ def main(
...
+    if stream:
+        for generation in response:
+            print(generation.generations[0].text, end="", flush=True)
+        print()
+    else:
+        print(response.generations[0].text)
 
 
-if __name__ == "__main__":
+if "__main__" == __name__:
```
Comment (Satrat) [inline src/deepsparse/transformers/inference/infer.py:246]: I recognize these are equivalent but it feels so wrong to see this comparison in the other direction....

**Review by Satrat — APPROVED** [review]

File: src/deepsparse/transformers/inference/infer.py (hunk 4 above)
Comment (rahul-tuli) [reply src/deepsparse/transformers/inference/infer.py:246]: Aah, leaving this as is for now to get the diff in, will update in a follow up!

**Pull request merged**